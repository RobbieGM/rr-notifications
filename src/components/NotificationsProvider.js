import React, { createContext, useState } from 'react';
import PropTypes from 'prop-types';
import Container from './Container';
import NotificationContainer from './NotificationContainer';

export const NotificationsContext = createContext({});

const propTypes = {
  children: PropTypes.node.isRequired,
  renderNotification: PropTypes.func.isRequired,
  position: PropTypes.array,
  stackNextOn: PropTypes.string,
  animationDuration: PropTypes.number,
  animationEasing: PropTypes.string,
  dismissAfter: PropTypes.number,
  slideFromSide: PropTypes.string,
};

const defaultProps = {
  position: ['40px', '40px', 'auto', 'auto'],
  stackNextOn: 'top',
  animationDuration: 400,
  animationEasing: 'ease',
  dismissAfter: 10000,
  slideFromSide: '',
};

const Notifications = ({
  children,
  renderNotification,
  position,
  stackNextOn,
  animationDuration,
  animationEasing,
  slideFromSide,
  dismissAfter,
}) => {
  const [notifications, updateNotifications] = useState([]);
  const showNotification = (payload = {}) => updateNotifications(state => (
    state.concat({
      id: Date.now(),
      isVisible: true,
      payload,
    })
  ));
  const hideNotification = id => updateNotifications(state => (
    state.map(notification => (
      notification.id === id
        ? {
          ...notification,
          isVisible: false,
        }
        : notification
    ))
  ));
  const unmountNotification = id => updateNotifications(state => (
    state.filter(notification => notification.id !== id)
  ));
  const removeNotification = id => () => {
    hideNotification(id);
    setTimeout(() => unmountNotification(id), 10000);
  };

  return (
    <NotificationsContext.Provider value={{ showNotification, removeNotification }}>
      <Container
        position={position}
        stackNextOn={stackNextOn}
        animationDuration={animationDuration}
        animationEasing={animationEasing}
        slideFromSide={slideFromSide}
      >
        {notifications.map(({ id, payload, isVisible }) => (
          <NotificationContainer
            animatedMargin={position[0] === 'auto' ? 'bottom' : 'top'}
            key={id}
            id={id}
            animationDuration={animationDuration}
            animationEasing={animationEasing}
            slideFromSide={slideFromSide}
            dismissAfter={dismissAfter}
            isVisible={isVisible}
          >
            {renderNotification({
              removeNotification: removeNotification(id),
              payload,
            })}
          </NotificationContainer>
        ))}
      </Container>
      {children}
    </NotificationsContext.Provider>
  );
};

Notifications.propTypes = propTypes;
Notifications.defaultProps = defaultProps;

export default Notifications;
