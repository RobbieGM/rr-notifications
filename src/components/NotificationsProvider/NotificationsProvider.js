import React, { createContext, useState, useRef } from 'react';
import PropTypes from 'prop-types';
import Container from '../Container/Container';
import NotificationContainer from '../NotificationContainer/NotificationContainer';

export const NotificationsContext = createContext({});

const propTypes = {
  /** Children */
  children: PropTypes.node.isRequired,
  /** Render prop which passes down removeNotification function and notification payload */
  renderNotification: PropTypes.func.isRequired,
  /** Fixed position where all notifications are displayed */
  position: PropTypes.array,
  /** Duration of the show and hide animations in miliseconds */
  animationDuration: PropTypes.number,
  /** Animation timing function / cubic-bezier */
  animationEasing: PropTypes.string,
  /** Time in miliseconds after which the notification is automatically dismissed */
  dismissAfter: PropTypes.number,
  /**
   * Horizontal direction which notification appears from. If not provided,
   * the notification will appear from top or bottom, depending on position prop
   * */
  slideFromSide: PropTypes.string,
};

const defaultProps = {
  position: ['40px', '40px', 'auto', 'auto'],
  animationDuration: 400,
  animationEasing: 'ease',
  dismissAfter: 6000,
  slideFromSide: '',
};

const NotificationsProvider = ({
  children,
  renderNotification,
  position,
  animationDuration,
  animationEasing,
  slideFromSide,
  dismissAfter,
}) => {
  const timers = useRef({});

  const [notifications, updateNotifications] = useState({});

  const hideNotification = id => updateNotifications(state => ({
    ...state,
    [id]: {
      ...state[id],
      isVisible: false,
    },
  }));

  const unmountNotification = id => updateNotifications(state => (
    Object.keys(state)
      .filter(notificationId => notificationId !== id)
      .reduce((acc, notificationId) => ({
        ...acc,
        [notificationId]: state[notificationId],
      }), {})
  ));

  const removeNotification = id => () => {
    clearTimeout(timers.current[id]);
    hideNotification(id);
    setTimeout(() => unmountNotification(id), animationDuration);
  };

  const showNotification = (payload = {}) => () => updateNotifications((state) => {
    const id = Date.now().toString();
    timers.current[id] = setTimeout(() => {
      removeNotification(id)();
    }, dismissAfter);
    return {
      ...state,
      [id]: {
        id,
        isVisible: true,
        payload,
      },
    };
  });

  return (
    <NotificationsContext.Provider value={{ showNotification, removeNotification }}>
      <Container position={position}>
        {Object.values(notifications).map(({ id, payload, isVisible }) => (
          <NotificationContainer
            key={id}
            position={position}
            animationDuration={animationDuration}
            animationEasing={animationEasing}
            slideFromSide={slideFromSide}
            dismissAfter={dismissAfter}
            isVisible={isVisible}
          >
            {renderNotification({
              removeNotification: removeNotification(id),
              payload,
            })}
          </NotificationContainer>
        ))}
      </Container>
      {children}
    </NotificationsContext.Provider>
  );
};

NotificationsProvider.propTypes = propTypes;
NotificationsProvider.defaultProps = defaultProps;

export default NotificationsProvider;
