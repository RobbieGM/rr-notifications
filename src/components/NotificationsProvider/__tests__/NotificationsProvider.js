import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import { render, cleanup, fireEvent } from 'react-testing-library';
import 'jest-dom/extend-expect';

import NotificationsProvider, { NotificationsContext } from '../NotificationsProvider';

const TestChild = () => {
  const { showNotification } = useContext(NotificationsContext);
  return (
    <button type="button" onClick={showNotification}>
      show notification
    </button>
  );
};

const children = <TestChild />;

const Notification = ({ removeNotification }) => (
  <div>
    <button type="button" onClick={removeNotification}>
      notification
    </button>
  </div>
);

Notification.propTypes = {
  removeNotification: PropTypes.func.isRequired,
};

const requiredProps = {
  renderNotification: Notification,
};

const optionalProps = {
  position: ['40px', '40px', 'auto', 'auto'],
  animationDuration: 400,
  animationEasing: 'ease',
};

const renderComponent = (props = {}) => render(
  <NotificationsProvider {...requiredProps} {...props}>
    {children}
  </NotificationsProvider>,
);

describe('<NotificationsProvider />', () => {
  afterEach(cleanup);

  it('should render correctly', () => {
    const { container: { firstChild } } = renderComponent();
    expect(firstChild).toBeDefined();
    expect(firstChild).toMatchSnapshot();
  });

  it('should have children', () => {
    const { container, getByText } = renderComponent();
    const childrenElement = getByText('show notification');
    expect(container).toContainElement(childrenElement);
  });

  it('should render correctly with custom props', () => {
    const { container: { firstChild } } = renderComponent(optionalProps);
    expect(firstChild).toMatchSnapshot();
  });

  it('should render two notifications and then close them', () => {
    const { container, getAllByText, getByText } = renderComponent();
    const button = getByText('show notification');
    fireEvent.click(button);
    fireEvent.click(button);

    const notifications = getAllByText('notification');
    expect(notifications.length).toEqual(2);
    expect(container).toMatchSnapshot();

    fireEvent.click(notifications[0]);
    expect(notifications.length).toEqual(1);
    expect(container).toMatchSnapshot();

    fireEvent.click(notifications[0]);
    expect(notifications.length).toEqual(0);
    expect(container).toMatchSnapshot();
  });
});
