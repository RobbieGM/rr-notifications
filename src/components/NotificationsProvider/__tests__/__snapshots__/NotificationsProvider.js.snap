// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NotificationsProvider /> should render correctly 1`] = `
<div
  class="sc-bdVaJa bPYOLZ"
/>
`;

exports[`<NotificationsProvider /> should render correctly with custom props 1`] = `
<div
  class="sc-bdVaJa bPYOLZ"
/>
`;

exports[`<NotificationsProvider /> should render two notifications and then close them 1`] = `
<div>
  <div
    class="sc-bdVaJa bPYOLZ"
  >
    <div
      class="sc-bwzfXH fsdYYp"
    >
      <div>
        <button
          type="button"
        >
          notification
        </button>
      </div>
    </div>
    <div
      class="sc-bwzfXH fsdYYp"
    >
      <div>
        <button
          type="button"
        >
          notification
        </button>
      </div>
    </div>
  </div>
  <button
    type="button"
  >
    show notification
  </button>
</div>
`;
